malcolm/modules/scanning/parts/scanrunnerpart.py

```diff
@@ -0,0 +1,498 @@
+            self.scans_completed.set_value)
+
+        self.scan_successes = NumberMeta(
+            "int64", "Successful scans",
+            tags=[Widget.TEXTUPDATE.tag()]
+        ).create_attribute_model()
+        registrar.add_attribute_model(
+            "scanSuccesses",
+            self.scan_successes,
+            self.scan_successes.set_value)
+
+        self.scan_failures = NumberMeta(
+            "int64", "Failed scans",
+            tags=[Widget.TEXTUPDATE.tag()]
+        ).create_attribute_model()
+        registrar.add_attribute_model(
+            "scanFailures",
+            self.scan_failures,
+            self.scan_failures.set_value)
+
+        self.output_directory = StringMeta(
+            "Root output directory (will create a sub-directory inside)",
+            tags=[config_tag(), Widget.TEXTINPUT.tag()]
+        ).create_attribute_model()
+        registrar.add_attribute_model(
+            "outputDirectory",
+            self.output_directory,
+            self.output_directory.set_value)
+
+        # Methods
+        registrar.add_method_model(self.loadFile)
+        registrar.add_method_model(self.run, needs_context=True)
+
+    def get_file_contents(self):
+        # type: () -> str
+
+        try:
+            with open(self.scan_file.value, "r") as input_file:
+                return input_file.read()
+        except IOError:
+            self.set_runner_state(RunnerStates.FAULT)
+            self.runner_status_message.set_value("Could not read scan file")
+            raise IOError("Could not read scan file")
+
+    def parse_yaml(self, string):
+        # type: (str) -> ...
+        try:
+            parsed_yaml = yaml.safe_load(string)
+            return parsed_yaml
+        except yaml.YAMLError:
+            self.set_runner_state(RunnerStates.FAULT)
+            self.runner_status_message.set_value("Could not parse scan file")
+            raise yaml.YAMLError("Could not parse scan file")
+
+    @staticmethod
+    def get_kwargs_from_dict(input_dict, kwargs_list):
+        kwargs = {}
+        if not isinstance(kwargs_list, list):
+            kwargs_list = [kwargs_list]
+        for kwarg in kwargs_list:
+            if kwarg in input_dict:
+                kwargs[kwarg] = input_dict[kwarg]
+        return kwargs
+
+    def parse_line_generator(self, entry):
+        # type: (dict) -> LineGenerator
+        axes = entry['axes']
+        units = entry['units']
+        start = entry['start']
+        stop = entry['stop']
+        size = entry['size']
+        kwargs = self.get_kwargs_from_dict(entry, 'alternate')
+
+        return LineGenerator(axes, units, start, stop, size, **kwargs)
+
+    def parse_compound_generator(self, entry):
+        # type: (dict) -> CompoundGenerator
+        duration = entry['duration']
+        generators = []
+        generators_dict = entry['generators']
+        for generator in generators_dict:
+            generators.append(self.parse_line_generator(generator['line']))
```

Comment:
```suggestion
            generators.append(LineGenerator.from_dict(generator['line']))
```
Should work...

Reply:
It did work, so I've committed the change.